nexus/src/app/certificate.rs

```diff
@@ -0,0 +1,230 @@
+            .db_datastore
+            .certificate_create(opctx, new_certificate)
+            .await?;
+        // TODO: Refresh other nexus servers?
+        self.refresh_tls_config(&opctx).await?;
+        info!(self.log, "TLS refreshed successfully");
+        Ok(cert)
+    }
+
+    pub async fn certificates_list(
+        &self,
+        opctx: &OpContext,
+        pagparams: &DataPageParams<'_, Name>,
+    ) -> ListResultVec<db::model::Certificate> {
+        self.db_datastore
+            .certificate_list_for(
+                opctx,
+                db::model::ServiceKind::Nexus,
+                pagparams,
+            )
+            .await
+    }
+
+    pub async fn certificate_delete(
+        self: &Arc<Self>,
+        opctx: &OpContext,
+        certificate_lookup: lookup::Certificate<'_>,
+    ) -> DeleteResult {
+        let (.., authz_cert, _db_cert) = certificate_lookup.fetch().await?;
+        self.db_datastore.certificate_delete(opctx, &authz_cert).await?;
+        // TODO: Refresh other nexus servers?
+        self.refresh_tls_config(&opctx).await?;
+        Ok(())
+    }
+
+    // Helper functions used by Nexus when managing its own server.
+
+    /// Returns the dropshot TLS configuration to run the Nexus external server.
+    pub async fn get_nexus_tls_config(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<Option<dropshot::ConfigTls>, Error> {
+        // Lookup x509 certificates which might be stored in CRDB, specifically
+        // for launching the Nexus service.
+        //
+        // We only grab one certificate (see: the "limit" argument) because
+        // we're currently fine just using whatever certificate happens to be
+        // available (as long as it's for Nexus).
+        let certs = self
+            .datastore()
+            .certificate_list_for(
+                &opctx,
+                db::model::ServiceKind::Nexus,
+                &DataPageParams {
+                    marker: None,
+                    direction: dropshot::PaginationOrder::Ascending,
+                    limit: std::num::NonZeroU32::new(1).unwrap(),
+                },
+            )
+            .await
+            .map_err(|e| {
+                Error::internal_error(&format!(
+                    "Failed to list certificates: {e}"
+                ))
+            })?;
+
+        let certificate = if let Some(certificate) = certs.get(0) {
+            certificate
+        } else {
+            return Ok(None);
+        };
+
+        Ok(Some(dropshot::ConfigTls::AsBytes {
+            certs: certificate.cert.clone(),
+            key: certificate.key.clone(),
+        }))
+    }
+
+    /// Refreshes the TLS configuration for the currently-running Nexus external
+    /// server. This involves either:
+    /// - Creating a new HTTPS server if once does not exist
+    /// - Refreshing an existing HTTPS server if it already exists
+    /// - Tearing down an HTTPS server if no certificates exist
```

Comment:
Do we ever actually want to do this?

More generally I'm a little unclear about how we want this sort of management to work.  Right now it looks like there's no way to add a certificate that is not immediately used.  I wonder if it will be important to separate these steps.  Along these lines, if we remove a cert as soon as it's deleted, and someone wants to replace the cert by delete + create, might they accidentally brick the rack by doing it in the wrong order?

Reply:
I was trying to keep this PR simple - my goal was "simple CRUD API, and we'll use whatever certificate we have for Nexus".

Do you want me to add an explicit endpoint for "certificate refresh" and "certificate promotion / activation"? I could make something more complex, where we add endpoints using the CRUD API, but then can explicitly nominate one for usage? I do think that has the same issue that "eventually, you make a call to the server, where on success it uses a new certificate for future requests", unless we also want to add a timeout or multi-server setup?

Alternatively, I can go back up to dropshot, and add support for handling multiple x.509 certificates simultaneously? That might require some invasive changes to rustls though